### ISSUE TITLE
There is a discrepancy in the loss calculation between the paper and the code.

### ISSUE THREAD

**Issue by mocuishle-gdy** [swt-user/DMPO#3]
thank you for such a great approach!

It appears that the implementation in the code at [line 915 of dmpo_trainer_efficient.py](https://github.com/swt-user/DMPO/blob/main/fastchat/train/dmpo_trainer_efficient.py#L915) introduces an additional gamma^j term compared to the final loss calculation described in the paper.
is this expected?

**Comment by mocushile-gif** [comment]
`def get_batch_loss_metrics(
        self,
        model,
        batch: Dict[str, Union[List, torch.LongTensor]],
        train_eval: Literal["train", "eval"] = "train",
    ):
        metrics = {}

        batch['chosen_coefficient'] = torch.ones_like(batch['chosen_labels']).type(torch.FloatTensor)
        for i, index_list in enumerate(batch['chosen_index']):
            num_round = len(index_list)//2
            for j in range(num_round):
                start = index_list[2*j]
                end = index_list[2*j+1]
                if self.gamma<1:
                    alpha = self.gamma**j * (1-self.gamma**(num_round-j)) /  (1-self.gamma**(num_round))
                elif self.gamma==1:
                    alpha = (num_round-j)/num_round
                batch['chosen_coefficient'][i][start:end] = alpha
                
        batch['rejected_coefficient'] = torch.ones_like(batch['rejected_labels']).type(torch.FloatTensor)
        for i, index_list in enumerate(batch['rejected_index']):
            num_round = len(index_list)//2
            for j in range(num_round):
                start = index_list[2*j]
                end = index_list[2*j+1]
                if self.gamma<1:
                    alpha = self.gamma**j * (1-self.gamma**(num_round-j)) /  (1-self.gamma**(num_round))
                elif self.gamma==1:
                    alpha = (num_round-j)/num_round
                batch['rejected_coefficient'][i][start:end] = alpha`


**alpha = self.gamma**j * (1-self.gamma**(num_round-j)) /  (1-self.gamma**(num_round))** introduces an additional gamma^j term.
Is this discrepancy between the final implementation and the theoretical framework in the paper an expected difference?

**Comment by swt-user** [comment]
Thanks for pointing out this! 

The code is correct. And there is a typo of the discount function in the paper. The correct format should be $\phi(t,T)=\gamma^t \cdot (1-\gamma^{T-t})/(1-\gamma^T)$.  The Corollary and observation in the article are not affected. We will update the typos in the paper.

**Comment by mocuishle-gdy** [comment]
Thank you very much for your response. I have another question: In Step 1 of the training script, different tasks (such as alfworld, sciworld, webshop) are distinguished for SFT. How is the training order for these tasks arranged during the SFT process? Or are all the tasks trained in a mixed order?

**Comment by swt-user** [comment]
Different tasks are fine-tuned separately with SFT and DMPO.
